### ISSUE TITLE
Broken include paths when building without submodules

### ISSUE THREAD

**Issue by NexAdn** [uowuo/abaddon#35]
Me again :D 

I've tried extracting the vendored dependencies and I had luck extracting IXWebsocket.
However, when I tried using an external SimpleIni, I got these compiler errors:

```
$ LANG=c ninja
[1/65] Building CXX object CMakeFiles/abaddon.dir/abaddon.cpp.o
FAILED: CMakeFiles/abaddon.dir/abaddon.cpp.o 
/usr/bin/c++  -I/usr/include/simpleini -I. -I/usr/include/gtkmm-3.0 -I/usr/lib64/gtkmm-3.0/include -I/usr/include/gdkmm-3.0 -I/usr/lib64/gdkmm-3.0/include -I/usr/include/gtk-3.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/glibmm-2.4 -I/usr/lib64/glibmm-2.4/include -I/usr/include/giomm-2.4 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/lib64/giomm-2.4/include -I/usr/include/pangomm-1.4 -I/usr/lib64/pangomm-1.4/include -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/cairo -I/usr/include/freetype2 -I/usr/include/cairomm-1.0 -I/usr/lib64/cairomm-1.0/include -I/usr/include/atkmm-1.6 -I/usr/lib64/atkmm-1.6/include -I/usr/include/atk-1.0 -I/usr/include/sigc++-2.0 -I/usr/lib64/sigc++-2.0/include -O3 -DNDEBUG -std=gnu++17 -MD -MT CMakeFiles/abaddon.dir/abaddon.cpp.o -MF CMakeFiles/abaddon.dir/abaddon.cpp.o.d -o CMakeFiles/abaddon.dir/abaddon.cpp.o -c ../abaddon.cpp
In file included from ../abaddon.cpp:13:
../dialogs/verificationgate.hpp:4:10: fatal error: ../../discord/objects.hpp: No such file or directory
    4 | #include "../../discord/objects.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

What is interesting is that a file in `.../dialogs` wants to include a file in `.../discord`, but has `../../discord` in its `#include` statement.

This only happens if the build configuration is patched not to use the submodules. The relevant change is:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 82d405f..d0e775e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,7 +23,8 @@ if (NOT IXWEBSOCKET_LIBRARY)
        include_directories(IXWEBSOCKET_INCLUDE_DIRS)
 endif()
 
-include_directories(thirdparty/simpleini)
+#include_directories(thirdparty/simpleini)
+include_directories("/usr/include/simpleini")
 
 if(MINGW OR WIN32)
        link_libraries(ws2_32)

```

Thus, `include_directories(thirdparty/simpleini)` not only makes the submodule files available in the build, but is also a requirement for building files unrelated to simpleini.

The workaround is simple – just add a random directory two levels below the repository root to the include path.
The fix, however, involves adjusting a bunch of files which I haven't identified yet. I just noticed – upon trying to fix `.../dialogs/verificationgate.hpp` – that there are multiple occurences.

I recommend just adding `${CMAKE_CURRENT_SOURCE_DIR}` and make _all_ include statements of project files (not vendored dependencies!) relative to the repository root. This avoids the hassle of figuring out what `../discord/xyz.hpp` and similar mean when reading through the source code, creates a uniform look of all `#include` statements (thus improving the code style as well!) and – of course – avoids errors with messed up include paths like above very unlikely, if not impossible.

So, it is not a bug in the way this software is usually built, but should still be considered. Vendoring dependencies – while comfortable for software developers – is not very welcome to some distro maintainers as it introduces the problem that not only the package maintainers, but also the software developers have to keep an eye on software releases (and thus potential security fixes!) for the vendored dependencies. After all, keeping an application with network access secure and up to date is very important.

**Comment by ouwou** [comment]
so i guess whats happening is that `include_directories` puts `thirdparty/simpleini` in the include list, and the `../../discord/foo` turns into `thirdparty/simpleini/../../discord/foo` which resolves to `discord/foo` which happens to work out? would correcting the 2 broken `#include`s not be sufficient to fix the issue here? i guess i just dont see how this issue is particular to dependencies/submodules other than thats just what happened to cause it to compile correctly despite the broken paths.
as for making all the includes relative to the base directory thats a good idea and might be something worth doing from now on. i guess that would just look like a `(target_)include_directories(${CMAKE_CURRENT_SOURCE_DIR})`? (as u can tell im not much of a CMake expert)

**Comment by NexAdn** [comment]
> so i guess whats happening is that include_directories puts thirdparty/simpleini in the include list, and the ../../discord/foo turns into thirdparty/simpleini/../../discord/foo which resolves to discord/foo which happens to work out? 

That's exactly what's happening here. I wouldn't have noticed it if there was just some other directory two directories below the repo root in the include path.

> would correcting the 2 broken #includes not be sufficient to fix the issue here?

would definitely be, but it isn't a very elegant solution (I hope I made that clear in my initial issue description).

>  i guess i just dont see how this issue is particular to dependencies/submodules other than thats just what happened to cause it to compile correctly despite the broken paths.

It is definitely not directly related to submodules, but it make a compilation fail when building from a source tarball (apart from vendored dependencies, of course).

> as for making all the includes relative to the base directory thats a good idea and might be something worth doing from now on. i guess that would just look like a (target_)include_directories(${CMAKE_CURRENT_SOURCE_DIR})? (as u can tell im not much of a CMake expert)

Yes, the approach I recommend is just including the source directory (in this case, `${CMAKE_CURRENT_SOURCE_DIR}`). In other projects (e.g. where all sources are below `src`) I would include that `src` directory.
As a rule of thumb, just try to make the `#include` statements look uniform and easily readable.
This usually just means including the source directory and (if present) a header directory (which doesn't exist in this case).

And I recommend not using `include_directories()` but rather `target_include_directories()` so the settings only apply to certain targets. But in simple CMake projects there isn't much of a difference, so it's more or less not important which one you use.

> also i guess re. your last point i should try and have the CMakeLists.txt look for simpleini and include as a submodule if its not found in the same kinda way i do ixwebsocket? or a find module

Well, since simpleini is not very distro maintainer friendly (it really wants to be included as a submodule), it is best to write a find module to look for the header file in common directories (standard include directory, e.g. `/usr/include`).

To make builds more independent of the particular build machine, it might also be better to enforce either vendored or system-wide dependencies with configure-time options. But that's not such a big issue in this case. I've just had experience with projects like polybar which check for software installed on the system at compile time and configure their project's features accordingly. This is, of course, easy to use, but is very intransparent if e.g. polybar gets installed before one of those hidden dependencies and the user is facing the problem that polybar misses support for this certain feature, because polybar didn't see the hidden dependency installed at compile time. Still, this issue doesn't apply here, so it's not that important.

To sum it up, here's my recommendation:

  * Make all `#include` statements for header files which reside in the repository relative to the repository root without `./` or `../`
  * Add `target_include_directories(... ${CMAKE_CURRENT_SOURCE_DIR})` in CMakeLists.txt
  * Write a find module for simpleini
  * Search for simpleini using the find module
  * (optionall) Add options to enforce either using vendored or system-wide dependencies
